server/src/main/java/org/apache/iotdb/db/metadata/storagegroup/SGMManager.java

```diff
@@ -0,0 +1,1765 @@
+
+  // the log file seriesPath
+  private String logFilePath;
+  private File logFile;
+  private MLogWriter logWriter;
+
+  private TimeseriesStatistics timeseriesStatistics = TimeseriesStatistics.getInstance();
+  private MTreeBelowSG mtree;
+  // device -> DeviceMNode
+  private LoadingCache<PartialPath, IMNode> mNodeCache;
+  private TagManager tagManager;
+
+  // region Interfaces and Implementation of MManager initialization、snapshot、recover and clear
+  public SGMManager() {
+    mtreeSnapshotInterval = config.getMtreeSnapshotInterval();
+    mtreeSnapshotThresholdTime = config.getMtreeSnapshotThresholdTime() * 1000L;
+
+    // do not write log when recover
+    isRecovering = true;
+
+    int cacheSize = config.getmManagerCacheSize();
+    mNodeCache =
+        Caffeine.newBuilder()
+            .maximumSize(cacheSize)
+            .build(
+                new com.github.benmanes.caffeine.cache.CacheLoader<PartialPath, IMNode>() {
+                  @Override
+                  public @Nullable IMNode load(@NonNull PartialPath partialPath)
+                      throws MetadataException {
+
+                    return mtree.getNodeByPath(partialPath);
+                  }
+                });
+
+    if (config.isEnableMTreeSnapshot()) {
+      timedCreateMTreeSnapshotThread =
+          IoTDBThreadPoolFactory.newSingleThreadScheduledExecutor("timedCreateMTreeSnapshot");
+      timedCreateMTreeSnapshotThread.scheduleAtFixedRate(
+          this::checkMTreeModified,
+          MTREE_SNAPSHOT_THREAD_CHECK_TIME,
+          MTREE_SNAPSHOT_THREAD_CHECK_TIME,
+          TimeUnit.SECONDS);
+    }
+
+    if (config.getSyncMlogPeriodInMs() != 0) {
+      timedForceMLogThread =
+          IoTDBThreadPoolFactory.newSingleThreadScheduledExecutor("timedForceMLogThread");
+
+      timedForceMLogThread.scheduleAtFixedRate(
+          this::forceMlog,
+          config.getSyncMlogPeriodInMs(),
+          config.getSyncMlogPeriodInMs(),
+          TimeUnit.MILLISECONDS);
+    }
+  }
+
+  // Because the writer will be used later and should not be closed here.
+  @SuppressWarnings("squid:S2093")
+  public synchronized void init(IStorageGroupMNode storageGroupMNode) {
+    if (initialized) {
+      return;
+    }
+
+    storageGroupFullPath = storageGroupMNode.getFullPath();
+
+    sgSchemaDirPath = config.getSchemaDir() + File.separator + storageGroupFullPath;
+    File sgSchemaFolder = SystemFileFactory.INSTANCE.getFile(sgSchemaDirPath);
+    if (!sgSchemaFolder.exists()) {
+      if (sgSchemaFolder.mkdirs()) {
+        logger.info("create storage group schema folder {}", sgSchemaFolder.getAbsolutePath());
+      } else {
+        logger.info(
+            "create storage group schema folder {} failed.", sgSchemaFolder.getAbsolutePath());
+      }
+    }
+    logFilePath = sgSchemaDirPath + File.separator + MetadataConstant.METADATA_LOG;
+
+    logFile = SystemFileFactory.INSTANCE.getFile(logFilePath);
+
+    try {
+      isRecovering = true;
```

Comment:
Maybe we need to consider combine the `init()` with constructor. `isRecovering = true;` appears both in constructor and `init()`, which is confusing. As general thinking, this line should appear only once in the whole procedure of initialization of SGMManager

Reply:
Thanks for suggestion. Already fixed.